src/commands/watch.rs

```diff
@@ -78,182 +73,177 @@ pub struct WatchCommand {
+            .context("Error creating artifact watcher")?;
+
+        let build_files_watcher_handle = if self.skip_build {
+            tokio::task::spawn(tokio::time::sleep(tokio::time::Duration::MAX))
+        } else {
+            let build_filterer = self
+                .build_filterer(&manifest_file, &manifest_dir, &manifest)
+                .await
+                .context("Error creating build files filterer")?;
+            self.spawn_watchexec(&manifest_dir, build_filterer, source_code_tx, "build")
+                .context("Error creating build files watcher")?
         };
-        spin_args.append(&mut vec![String::from("-f"), manifest_path]);
-        if matches!(state, State::Running) {
-            spin_args.extend(up_args);
-        }
-        spin_args
-    }
 
-    async fn generate_filter_config(&self) -> Result<crate::watch_filter::Config> {
-        let app = spin_common::paths::resolve_manifest_file_path(&self.app_source)?;
-        let app_dir = parent_dir(&app)?;
-        let app_manifest = spin_loader::local::raw_manifest_from_file(&app)
-            .await?
-            .into_v1();
+        // Start the the uppificator and buildifier to process notifications from the
+        // `watchexec` tasks.
 
-        // We always want to watch the application manifest
-        let manifest_pattern = WatchPattern::new(
-            app.to_str()
-                .with_context(|| format!("non-unicode manifest path {:?}", app))?
-                .to_owned(),
-            app_dir.as_path(),
-        )?;
-
-        // We want to watch the source code if we aren't skipping the build step
-        let source_patterns = match !self.skip_build {
-            false => vec![],
-            true => app_manifest
-                .components
-                .iter()
-                .filter_map(|c| WatchCommand::create_source_pattern(c, &app_dir))
-                .flatten()
-                .collect::<Result<Vec<WatchPattern>>>()?,
+        let uppificator_handle = tokio::task::spawn(async move { uppificator.run().await });
+        let buildifier_handle = if self.skip_build {
+            tokio::task::spawn(tokio::time::sleep(tokio::time::Duration::MAX))
+        } else {
+            tokio::task::spawn(async move { buildifier.run().await })
         };
 
-        // We want to watch a component's source if it has no build step. If we're skipping
-        // building then we'll watch all component sources.
-        let component_source_patterns = app_manifest
+        // Wait for either the user to stop the watch, or some catastrophe to dump us out.
+
+        _ = ctrlc::set_handler(move || {
+            _ = stop_tx.send(Uuid::new_v4());
+        });
+
+        _ = futures::future::select_all(vec![
+            uppificator_handle,
+            buildifier_handle,
+            artifact_watcher_handle,
+            build_files_watcher_handle,
+        ])
+        .await;
+
+        Ok(())
+    }
+
+    fn spawn_watchexec(
```

Comment:
In terms of system notify limits you mean?  That's a reasonable concern for sure, and one I probably wouldn't pick up in simple test cases.  I did try it on the docs site but that's still medium-scale rather than enormous.  The default notify limit is, I think, 8192.  What do you reckon?

Reply:
Honestly I meant in terms of other system resources like CPU, memory which is probably far from any real concern. Good point on the notify limits. I'd say we're probably handling that when it actually becomes a problem.